copyStorePath: Fix hash errors when copying from older store

FPtje opened the pull request:
This commit partially reverts 48662d151bdf4a38670897beacea9d1bd750376a. When
copying from an older store (in my case a store running Nix 1.11.7), nix would
throw errors about there being no hash. This is fixed by recalculating the hash.

From the Nixcon 2018 hackathon

edolstra commented on src/libstore/store-api.cc:
```diff
@@ -589,14 +589,29 @@ void copyStorePath(ref<Store> srcStore, ref<Store> dstStore,
+            };
+        };
+        MyStringSink sink(progress);
+        srcStore->narFromPath({storePath}, sink);
```
Note that this will cause the NAR to be copied twice (due to the later call to `narFromPath`). You could create a `StringSource` for `sink.s` and pass it to `dstStore->addToStore()`.

BTW I think the whole progress stuff can be removed here since this is just a backward compatibility path.

FPtje commented:
I've force pushed the following changes to the code:

- Fixed NAR being copied twice by indeed creating a `StringSource` for `sink.s` and passing it to `dstStore->addToStore()`. Thanks for helping me along the way there. I'm not familiar with the Nix code base, and your hint sent me along the right path.
- Removed the progress stuff.
- Removed `FIXME` comment, as it now seems to be properly fixed

I've tested by copying things to and from a Nix 1.11.7 store and to and from a nix 2.0.4 store.

edolstra commented:
Thanks!
